File: goldenmatch/core/probabilistic.py
"""Fellegi-Sunter probabilistic matching with EM-trained parameters.

Implements the classic Fellegi-Sunter model for record linkage:
- Comparison vectors classify field agreements into levels (agree/partial/disagree)
- Expectation-Maximization estimates m-probabilities (P(level|match)) and
  u-probabilities (P(level|non-match)) from unlabeled data
- Match weights are log-likelihood ratios: log2(m/u)
- Thresholds computed from the weight distribution

References:
    Fellegi & Sunter (1969). "A Theory for Record Linkage"
    Winkler (2006). "Overview of Record Linkage and Current Research Directions"
"""
from __future__ import annotations

import logging
import math
import random
from dataclasses import dataclass, field
from itertools import combinations

import numpy as np
import polars as pl

from goldenmatch.config.schemas import MatchkeyConfig
from goldenmatch.core.scorer import score_field

logger = logging.getLogger(__name__)


@dataclass
class EMResult:
    """Result of EM training for Fellegi-Sunter model."""

    m_probs: dict[str, list[float]]  # field -> P(level_i | match)
    u_probs: dict[str, list[float]]  # field -> P(level_i | non-match)
    match_weights: dict[str, list[float]]  # field -> log2(m/u) per level
    converged: bool
    iterations: int
    proportion_matched: float  # estimated match rate in the data


def comparison_vector(
    row_a: dict,
    row_b: dict,
    mk: MatchkeyConfig,
) -> list[int]:
    """Compute comparison vector for a pair of records.

    Returns a list of level indices, one per field.
    For 2-level fields: 0=disagree, 1=agree
    For 3-level fields: 0=disagree, 1=partial, 2=agree
    """
    levels = []
    for f in mk.fields:
        val_a = str(row_a.get(f.field, "")) if row_a.get(f.field) is not None else None
        val_b = str(row_b.get(f.field, "")) if row_b.get(f.field) is not None else None
        s = score_field(val_a, val_b, f.scorer)

        if s is None:
            levels.append(0)  # treat nulls as disagree
        elif f.levels == 2:
            levels.append(1 if s >= f.partial_threshold else 0)
        else:  # 3 levels
            if s >= 0.95:
                levels.append(2)  # agree
            elif s >= f.partial_threshold:
                levels.append(1)  # partial
            else:
                levels.append(0)  # disagree
    return levels


def _sample_pairs(
    df: pl.DataFrame,
    n_pairs: int = 10000,
    seed: int = 42,
) -> list[tuple[int, int]]:
    """Sample random pairs for EM training."""
    row_ids = df["__row_id__"].to_list()
    rng = random.Random(seed)

    if len(row_ids) < 2:
        return []

    # For small datasets, use all pairs
    max_possible = len(row_ids) * (len(row_ids) - 1) // 2
    if max_possible <= n_pairs:
        return list(combinations(row_ids, 2))

    # Reservoir sampling of random pairs
    pairs = set()
    attempts = 0
    max_attempts = n_pairs * 10
    while len(pairs) < n_pairs and attempts < max_attempts:
        i, j = rng.sample(row_ids, 2)
        pair = (min(i, j), max(i, j))
        pairs.add(pair)
        attempts += 1

    return list(pairs)


def _build_comparison_matrix(
    pairs: list[tuple[int, int]],
    row_lookup: dict[int, dict],
    mk: MatchkeyConfig,
) -> np.ndarray:
    """Build NxF comparison matrix where N=pairs, F=fields."""
    n_pairs = len(pairs)
    n_fields = len(mk.fields)
    matrix = np.zeros((n_pairs, n_fields), dtype=np.int8)

    for i, (a, b) in enumerate(pairs):
        row_a = row_lookup.get(a, {})
        row_b = row_lookup.get(b, {})
        vec = comparison_vector(row_a, row_b, mk)
        matrix[i] = vec

    return matrix


def train_em(
    df: pl.DataFrame,
    mk: MatchkeyConfig,
    n_sample_pairs: int = 10000,
    max_iterations: int = 20,
    convergence: float = 0.001,
    seed: int = 42,
) -> EMResult:
    """Train Fellegi-Sunter model using Expectation-Maximization.

    Args:
        df: DataFrame with __row_id__ and field columns.
        mk: Probabilistic matchkey config.
        n_sample_pairs: Number of random pairs to sample for training.
        max_iterations: Maximum EM iterations.
        convergence: Stop when max change in any probability < this.
        seed: Random seed for pair sampling.

    Returns:
        EMResult with trained m/u probabilities and match weights.
    """
    cols = [f.field for f in mk.fields if f.field != "__record__"]
    row_lookup: dict[int, dict] = {}
    for row in df.select(["__row_id__"] + cols).to_dicts():
        row_lookup[row["__row_id__"]] = row

    # Sample pairs
    pairs = _sample_pairs(df, n_sample_pairs, seed)
    if len(pairs) < 10:
        logger.warning("Too few pairs (%d) for EM training", len(pairs))
        return _fallback_result(mk)

    # Build comparison matrix
    comp_matrix = _build_comparison_matrix(pairs, row_lookup, mk)
    n_pairs = len(pairs)
    n_fields = len(mk.fields)

    # Initialize parameters
    p_match = 0.05  # prior: 5% of pairs are matches
    # Initialize m/u with reasonable priors
    m_probs = {}
    u_probs = {}
    for j, f in enumerate(mk.fields):
        n_levels = f.levels
        if n_levels == 2:
            m_probs[f.field] = [0.1, 0.9]  # matches mostly agree
            u_probs[f.field] = [0.9, 0.1]  # non-matches mostly disagree
        else:
            m_probs[f.field] = [0.05, 0.15, 0.80]  # matches: mostly full agree
            u_probs[f.field] = [0.80, 0.15, 0.05]  # non-matches: mostly disagree

    # EM iterations
    converged = False
    for iteration in range(max_iterations):
        old_m = {k: list(v) for k, v in m_probs.items()}
        old_u = {k: list(v) for k, v in u_probs.items()}

        # E-step: compute posterior P(match | comparison vector)
        posteriors = np.zeros(n_pairs)
        for i in range(n_pairs):
            log_m = 0.0
            log_u = 0.0
            for j, f in enumerate(mk.fields):
                level = comp_matrix[i, j]
                # Clamp to avoid log(0)
                m_val = max(m_probs[f.field][level], 1e-10)
                u_val = max(u_probs[f.field][level], 1e-10)
                log_m += math.log(m_val)
                log_u += math.log(u_val)

            log_match = math.log(max(p_match, 1e-10)) + log_m
            log_nonmatch = math.log(max(1 - p_match, 1e-10)) + log_u

            # Numerically stable softmax
            max_log = max(log_match, log_nonmatch)
            posteriors[i] = math.exp(log_match - max_log) / (
                math.exp(log_match - max_log) + math.exp(log_nonmatch - max_log)
            )

        # M-step: update parameters
        total_match = posteriors.sum()
        total_nonmatch = n_pairs - total_match
        p_match = max(total_match / n_pairs, 1e-6)

        for j, f in enumerate(mk.fields):
            n_levels = f.levels
            new_m = [0.0] * n_levels
            new_u = [0.0] * n_levels

            for level in range(n_levels):
                mask = comp_matrix[:, j] == level
                new_m[level] = (posteriors[mask].sum() + 1e-6) / (total_match + n_levels * 1e-6)
                new_u[level] = ((1 - posteriors[mask]).sum() + 1e-6) / (total_nonmatch + n_levels * 1e-6)

            m_probs[f.field] = new_m
            u_probs[f.field] = new_u

        # Check convergence
        max_delta = 0.0
        for f in mk.fields:
            for k in range(f.levels):
                max_delta = max(max_delta, abs(m_probs[f.field][k] - old_m[f.field][k]))
                max_delta = max(max_delta, abs(u_probs[f.field][k] - old_u[f.field][k]))

        if max_delta < convergence:
            converged = True
            logger.info("EM converged after %d iterations (delta=%.6f)", iteration + 1, max_delta)
            break

    if not converged:
        logger.warning("EM did not converge after %d iterations (delta=%.6f)", max_iterations, max_delta)

    # Compute match weights: log2(m/u)
    match_weights = {}
    for f in mk.fields:
        weights = []
        for k in range(f.levels):
            m_val = max(m_probs[f.field][k], 1e-10)
            u_val = max(u_probs[f.field][k], 1e-10)
            weights.append(math.log2(m_val / u_val))
        match_weights[f.field] = weights

    return EMResult(
        m_probs=m_probs,
        u_probs=u_probs,
        match_weights=match_weights,
        converged=converged,
        iterations=min(iteration + 1, max_iterations) if not converged else iteration + 1,
        proportion_matched=p_match,
    )


def _fallback_result(mk: MatchkeyConfig) -> EMResult:
    """Return a fallback EMResult when EM can't be trained."""
    m_probs = {}
    u_probs = {}
    match_weights = {}
    for f in mk.fields:
        if f.levels == 2:
            m_probs[f.field] = [0.1, 0.9]
            u_probs[f.field] = [0.9, 0.1]
            match_weights[f.field] = [math.log2(0.1 / 0.9), math.log2(0.9 / 0.1)]
        else:
            m_probs[f.field] = [0.05, 0.15, 0.80]
            u_probs[f.field] = [0.80, 0.15, 0.05]
            match_weights[f.field] = [
                math.log2(0.05 / 0.80),
                math.log2(0.15 / 0.15),
                math.log2(0.80 / 0.05),
            ]
    return EMResult(
        m_probs=m_probs, u_probs=u_probs, match_weights=match_weights,
        converged=False, iterations=0, proportion_matched=0.05,
    )


def compute_thresholds(em_result: EMResult) -> tuple[float, float]:
    """Compute link and review thresholds from EM result.

    Returns (link_threshold, review_threshold) as normalized 0-1 scores.
    link_threshold: pairs above this are matches
    review_threshold: pairs between review and link are uncertain
    """
    # Compute max possible weight (all fields agree at highest level)
    max_weight = 0.0
    min_weight = 0.0
    for field_name, weights in em_result.match_weights.items():
        max_weight += max(weights)
        min_weight += min(weights)

    weight_range = max_weight - min_weight
    if weight_range == 0:
        return 0.85, 0.60

    # Link threshold at ~85th percentile of weight range
    link_raw = min_weight + 0.85 * weight_range
    # Review threshold at ~60th percentile
    review_raw = min_weight + 0.60 * weight_range

    # Normalize to 0-1
    link_norm = (link_raw - min_weight) / weight_range
    review_norm = (review_raw - min_weight) / weight_range

    return round(link_norm, 4), round(review_norm, 4)


def score_probabilistic(
    block_df: pl.DataFrame,
    mk: MatchkeyConfig,
    em_result: EMResult,
    exclude_pairs: set[tuple[int, int]] | None = None,
) -> list[tuple[int, int, float]]:
    """Score pairs in a block using Fellegi-Sunter match weights.

    Returns pairs above the link threshold as (row_id_a, row_id_b, normalized_score).
    Score is normalized to 0-1 range for compatibility with the rest of the pipeline.
    """
    if exclude_pairs is None:
        exclude_pairs = set()

    # Build row lookup
    cols = [f.field for f in mk.fields if f.field != "__record__"]
    row_lookup: dict[int, dict] = {}
    for row in block_df.select(["__row_id__"] + cols).to_dicts():
        row_lookup[row["__row_id__"]] = row

    row_ids = block_df["__row_id__"].to_list()

    # Compute weight range for normalization
    max_weight = sum(max(em_result.match_weights[f.field]) for f in mk.fields)
    min_weight = sum(min(em_result.match_weights[f.field]) for f in mk.fields)
    weight_range = max_weight - min_weight

    # Determine threshold
    if mk.link_threshold is not None:
        link_threshold = mk.link_threshold
    else:
        link_threshold, _ = compute_thresholds(em_result)

    results = []
    for i in range(len(row_ids)):
        for j in range(i + 1, len(row_ids)):
            a, b = row_ids[i], row_ids[j]
            pair_key = (min(a, b), max(a, b))
            if pair_key in exclude_pairs:
                continue

            row_a = row_lookup.get(a, {})
            row_b = row_lookup.get(b, {})
            vec = comparison_vector(row_a, row_b, mk)

            # Sum match weights
            total_weight = 0.0
            for k, f in enumerate(mk.fields):
                total_weight += em_result.match_weights[f.field][vec[k]]

            # Normalize to 0-1
            if weight_range > 0:
                normalized = (total_weight - min_weight) / weight_range
            else:
                normalized = 0.5

            if normalized >= link_threshold:
                results.append((a, b, round(normalized, 4)))

    return results


def score_pair_probabilistic(
    row_a: dict,
    row_b: dict,
    mk: MatchkeyConfig,
    em_result: EMResult,
) -> float:
    """Score a single pair using Fellegi-Sunter weights. For match_one."""
    vec = comparison_vector(row_a, row_b, mk)

    max_weight = sum(max(em_result.match_weights[f.field]) for f in mk.fields)
    min_weight = sum(min(em_result.match_weights[f.field]) for f in mk.fields)
    weight_range = max_weight - min_weight

    total_weight = 0.0
    for k, f in enumerate(mk.fields):
        total_weight += em_result.match_weights[f.field][vec[k]]

    if weight_range > 0:
        return (total_weight - min_weight) / weight_range
    return 0.5

File: packages/goldenmatch-js/src/core/probabilistic.ts
/**
 * probabilistic.ts — Fellegi-Sunter probabilistic matching with EM-trained
 * parameters. Ports `goldenmatch/core/probabilistic.py` (discrete path).
 *
 * Implements:
 * - Comparison vectors (2/3/N-level field agreements)
 * - Splink-style EM: u estimated from random pairs (fixed), m trained via EM
 * - Blocking fields get fixed neutral priors
 * - Match weights as log2(m/u) log-likelihood ratios, normalized to [0,1]
 *
 * Edge-safe: no `node:` imports, no numpy. Uses typed arrays where helpful.
 */

import type { Row, MatchkeyConfig, MatchkeyField, ScoredPair } from "./types.js";
import { makeScoredPair } from "./types.js";
import { scoreField, asString } from "./scorer.js";
import { applyTransforms } from "./transforms.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface EMOptions {
  readonly maxIterations?: number;
  readonly convergence?: number;
  readonly blockingFields?: readonly string[];
  readonly seed?: number;
  readonly nSamplePairs?: number;
}

export interface EMResult {
  /** P(level | match) per field. */
  readonly m: Readonly<Record<string, readonly number[]>>;
  /** P(level | non-match) per field. */
  readonly u: Readonly<Record<string, readonly number[]>>;
  /** log2(m / u) per level per field. Score weights. */
  readonly matchWeights: Readonly<Record<string, readonly number[]>>;
  /** Estimated p(match) in the sampled population. */
  readonly proportionMatched: number;
  readonly iterations: number;
  readonly converged: boolean;
}

// ---------------------------------------------------------------------------
// Deterministic RNG (xorshift32) — avoids relying on Math.random's seedability
// ---------------------------------------------------------------------------

function makeRng(seed: number): () => number {
  let x = seed | 0 || 1;
  return () => {
    x ^= x << 13;
    x ^= x >>> 17;
    x ^= x << 5;
    // Return in [0, 1): divide by 2^32 (not 2^32-1) so the value cannot reach 1.0.
    return (x >>> 0) / 0x100000000;
  };
}

// ---------------------------------------------------------------------------
// Field levels helper
// ---------------------------------------------------------------------------

function fieldLevels(f: MatchkeyField): number {
  return f.levels ?? 2;
}

function fieldPartialThreshold(f: MatchkeyField): number {
  return f.partialThreshold ?? 0.7;
}

// ---------------------------------------------------------------------------
// Public: buildComparisonVector
// ---------------------------------------------------------------------------

/**
 * Build a comparison vector: one integer level per field.
 *   levels=2: 0=disagree, 1=agree
 *   levels=3: 0=disagree, 1=partial, 2=agree (>= 0.95)
 *   levels=N: evenly spaced thresholds k/N for k in 1..N-1
 */
export function buildComparisonVector(
  rowA: Row,
  rowB: Row,
  fields: readonly MatchkeyField[],
): readonly number[] {
  const levels: number[] = [];
  for (const f of fields) {
    let valA = asString(rowA[f.field]);
    let valB = asString(rowB[f.field]);
    if (f.transforms.length > 0) {
      valA = applyTransforms(valA, f.transforms);
      valB = applyTransforms(valB, f.transforms);
    }
    const s = scoreField(valA, valB, f.scorer);
    const n = fieldLevels(f);
    const partial = fieldPartialThreshold(f);

    if (s === null) {
      levels.push(0);
      continue;
    }

    if (n === 2) {
      levels.push(s >= partial ? 1 : 0);
    } else if (n === 3) {
      if (s >= 0.95) levels.push(2);
      else if (s >= partial) levels.push(1);
      else levels.push(0);
    } else {
      let level = 0;
      for (let k = 1; k < n; k++) {
        if (s >= k / n) level = k;
      }
      levels.push(level);
    }
  }
  return levels;
}

// ---------------------------------------------------------------------------
// Random-pair sampling (used for u estimation)
// ---------------------------------------------------------------------------

function samplePairs(
  rows: readonly Row[],
  nPairs: number,
  rand: () => number,
): Array<readonly [number, number]> {
  const ids: number[] = [];
  for (const r of rows) {
    const id = r["__row_id__"];
    if (typeof id === "number") ids.push(id);
  }
  if (ids.length < 2) return [];

  const maxPossible = (ids.length * (ids.length - 1)) / 2;
  if (maxPossible <= nPairs) {
    const out: Array<readonly [number, number]> = [];
    for (let i = 0; i < ids.length; i++) {
      for (let j = i + 1; j < ids.length; j++) {
        out.push([ids[i]!, ids[j]!] as const);
      }
    }
    return out;
  }

  const seen = new Set<string>();
  const pairs: Array<readonly [number, number]> = [];
  const maxAttempts = nPairs * 10;
  let attempts = 0;
  while (pairs.length < nPairs && attempts < maxAttempts) {
    attempts++;
    const i = Math.floor(rand() * ids.length);
    let j = Math.floor(rand() * ids.length);
    if (j === i) j = (j + 1) % ids.length;
    const a = Math.min(ids[i]!, ids[j]!);
    const b = Math.max(ids[i]!, ids[j]!);
    const key = `${a}:${b}`;
    if (seen.has(key)) continue;
    seen.add(key);
    pairs.push([a, b] as const);
  }
  return pairs;
}

function buildComparisonMatrix(
  pairs: ReadonlyArray<readonly [number, number]>,
  rowById: ReadonlyMap<number, Row>,
  fields: readonly MatchkeyField[],
): number[][] {
  const out: number[][] = [];
  for (const [a, b] of pairs) {
    const rowA = rowById.get(a) ?? {};
    const rowB = rowById.get(b) ?? {};
    const vec = buildComparisonVector(rowA, rowB, fields);
    out.push([...vec]);
  }
  return out;
}

// ---------------------------------------------------------------------------
// Public: trainEM
// ---------------------------------------------------------------------------

/**
 * Splink-style EM training:
 *   1. Estimate u from random pairs (fixed throughout).
 *   2. Train m via EM starting from exponential priors.
 *   3. Blocking fields bypass EM and receive fixed neutral u + linear weights.
 */
export function trainEM(
  rows: readonly Row[],
  mk: MatchkeyConfig,
  options?: EMOptions,
): EMResult {
  // Probabilistic-only parameters; fall through to defaults for other variants.
  const emIterations =
    mk.type === "probabilistic" ? mk.emIterations : undefined;
  const convergenceThreshold =
    mk.type === "probabilistic" ? mk.convergenceThreshold : undefined;
  const maxIterations = options?.maxIterations ?? emIterations ?? 20;
  const convergence = options?.convergence ?? convergenceThreshold ?? 0.001;
  const blockingFields = new Set(options?.blockingFields ?? []);
  const seed = options?.seed ?? 42;
  const nSamplePairs = options?.nSamplePairs ?? 10000;

  const fields = mk.fields;
  if (fields.length === 0) return fallbackResult(mk);

  const rand = makeRng(seed);
  const rowById = new Map<number, Row>();
  for (const r of rows) {
    const id = r["__row_id__"];
    if (typeof id === "number") rowById.set(id, r);
  }

  // Step 1: u from random pairs.
  const sampleForU = samplePairs(rows, Math.min(nSamplePairs, 5000), rand);
  if (sampleForU.length < 10) return fallbackResult(mk);
  const uMatrix = buildComparisonMatrix(sampleForU, rowById, fields);

  const u: Record<string, number[]> = {};
  fields.forEach((f, j) => {
    const n = fieldLevels(f);
    const counts = new Array<number>(n).fill(0);
    for (const row of uMatrix) {
      const lvl = row[j]!;
      if (lvl >= 0 && lvl < n) counts[lvl]! += 1;
    }
    const total = counts.reduce((a, b) => a + b, 0) + n * 1e-6;
    u[f.field] = counts.map((c) => (c + 1e-6) / total);
  });

  // Blocking fields get neutral u.
  for (const f of fields) {
    if (blockingFields.has(f.field)) {
      const n = fieldLevels(f);
      if (n === 2) u[f.field] = [0.5, 0.5];
      else u[f.field] = [0.34, 0.33, ...new Array<number>(n - 2).fill(0.33 / Math.max(1, n - 2))];
    }
  }

  // Step 2: m priors (exponential: highest level gets most mass).
  const m: Record<string, number[]> = {};
  for (const f of fields) {
    const n = fieldLevels(f);
    const raw: number[] = [];
    for (let k = 0; k < n; k++) raw.push(2 ** k);
    const sum = raw.reduce((a, b) => a + b, 0);
    m[f.field] = raw.map((r) => r / sum);
  }

  // Use the same random-pair matrix for EM. In Python, blocked pairs are
  // preferred when available; we don't have blocks in this entry point, so
  // we train on the random sample (the fallback path).
  const compMatrix = uMatrix;
  const nPairs = compMatrix.length;

  let pMatch = 0.02;
  let converged = false;
  let iterations = 0;

  for (let iter = 0; iter < maxIterations; iter++) {
    iterations = iter + 1;
    const oldM: Record<string, number[]> = {};
    for (const k of Object.keys(m)) oldM[k] = [...m[k]!];

    // E-step.
    const posteriors = new Float64Array(nPairs);
    for (let i = 0; i < nPairs; i++) {
      let logM = Math.log(Math.max(pMatch, 1e-10));
      let logU = Math.log(Math.max(1 - pMatch, 1e-10));
      for (let j = 0; j < fields.length; j++) {
        const f = fields[j]!;
        const level = compMatrix[i]![j]!;
        const mProb = Math.max(m[f.field]![level] ?? 1e-10, 1e-10);
        const uProb = Math.max(u[f.field]![level] ?? 1e-10, 1e-10);
        logM += Math.log(mProb);
        logU += Math.log(uProb);
      }
      const maxLog = Math.max(logM, logU);
      const em = Math.exp(logM - maxLog);
      const eu = Math.exp(logU - maxLog);
      posteriors[i] = em / (em + eu);
    }

    // M-step (m only).
    let totalMatch = 0;
    for (let i = 0; i < nPairs; i++) totalMatch += posteriors[i]!;
    pMatch = Math.max(totalMatch / nPairs, 1e-6);

    for (let j = 0; j < fields.length; j++) {
      const f = fields[j]!;
      if (blockingFields.has(f.field)) continue;
      const n = fieldLevels(f);
      const newM = new Array<number>(n).fill(0);
      for (let i = 0; i < nPairs; i++) {
        const level = compMatrix[i]![j]!;
        if (level >= 0 && level < n) newM[level]! += posteriors[i]!;
      }
      const denom = totalMatch + n * 1e-6;
      for (let k = 0; k < n; k++) {
        newM[k] = (newM[k]! + 1e-6) / denom;
      }
      m[f.field] = newM;
    }

    // Convergence.
    let maxDelta = 0;
    for (const f of fields) {
      if (blockingFields.has(f.field)) continue;
      const n = fieldLevels(f);
      for (let k = 0; k < n; k++) {
        const d = Math.abs(m[f.field]![k]! - oldM[f.field]![k]!);
        if (d > maxDelta) maxDelta = d;
      }
    }
    if (maxDelta < convergence) {
      converged = true;
      break;
    }
  }

  // Match weights = log2(m/u), with fixed linear weights for blocking fields.
  const matchWeights: Record<string, number[]> = {};
  for (const f of fields) {
    const n = fieldLevels(f);
    if (blockingFields.has(f.field)) {
      const w: number[] = [];
      for (let k = 0; k < n; k++) {
        w.push(n > 1 ? -3.0 + (6.0 * k) / (n - 1) : 3.0);
      }
      matchWeights[f.field] = w;
      continue;
    }
    const w: number[] = [];
    for (let k = 0; k < n; k++) {
      const mVal = Math.max(m[f.field]![k]!, 1e-10);
      const uVal = Math.max(u[f.field]![k]!, 1e-10);
      w.push(Math.log2(mVal / uVal));
    }
    matchWeights[f.field] = w;
  }

  return {
    m: m as Readonly<Record<string, readonly number[]>>,
    u: u as Readonly<Record<string, readonly number[]>>,
    matchWeights: matchWeights as Readonly<Record<string, readonly number[]>>,
    proportionMatched: pMatch,
    iterations,
    converged,
  };
}

// ---------------------------------------------------------------------------
// Public: scoreProbabilistic
// ---------------------------------------------------------------------------

export interface ProbScoreOptions {
  readonly excludePairs?: ReadonlySet<string>;
  readonly threshold?: number;
}

/**
 * Score all pairs in a block using F-S match weights.
 * Returns normalized scores in [0,1] (weight sum mapped to 0-1 via min/max).
 * Pairs below threshold are filtered out.
 */
export function scoreProbabilistic(
  rows: readonly Row[],
  mk: MatchkeyConfig,
  em: EMResult,
  options?: ProbScoreOptions,
): ScoredPair[] {
  const fields = mk.fields;
  if (fields.length === 0) return [];

  const excludePairs = options?.excludePairs ?? new Set<string>();
  const linkThreshold =
    mk.type === "probabilistic" ? mk.linkThreshold : undefined;
  const threshold = options?.threshold ?? linkThreshold ?? 0.5;

  // Min/max possible weight totals for normalization.
  let maxWeight = 0;
  let minWeight = 0;
  for (const f of fields) {
    const w = em.matchWeights[f.field];
    if (!w || w.length === 0) continue;
    maxWeight += Math.max(...w);
    minWeight += Math.min(...w);
  }
  const weightRange = maxWeight - minWeight;

  const rowIds: number[] = [];
  const rowLookup: Row[] = [];
  for (const r of rows) {
    const id = r["__row_id__"];
    if (typeof id === "number") {
      rowIds.push(id);
      rowLookup.push(r);
    }
  }

  const results: ScoredPair[] = [];
  for (let i = 0; i < rowIds.length; i++) {
    for (let j = i + 1; j < rowIds.length; j++) {
      const a = Math.min(rowIds[i]!, rowIds[j]!);
      const b = Math.max(rowIds[i]!, rowIds[j]!);
      const key = `${a}:${b}`;
      if (excludePairs.has(key)) continue;

      const vec = buildComparisonVector(rowLookup[i]!, rowLookup[j]!, fields);

      let total = 0;
      for (let k = 0; k < fields.length; k++) {
        const f = fields[k]!;
        const level = vec[k]!;
        const w = em.matchWeights[f.field];
        if (!w) continue;
        total += w[level] ?? 0;
      }

      const normalized =
        weightRange > 0 ? (total - minWeight) / weightRange : 0.5;

      if (normalized >= threshold) {
        results.push(
          makeScoredPair(a, b, Math.round(normalized * 10000) / 10000),
        );
      }
    }
  }
  return results;
}

// ---------------------------------------------------------------------------
// Public: scoreProbabilisticPair (single-pair variant for match_one use)
// ---------------------------------------------------------------------------

export function scoreProbabilisticPair(
  rowA: Row,
  rowB: Row,
  mk: MatchkeyConfig,
  em: EMResult,
): number {
  const fields = mk.fields;
  if (fields.length === 0) return 0.5;

  let maxWeight = 0;
  let minWeight = 0;
  for (const f of fields) {
    const w = em.matchWeights[f.field];
    if (!w || w.length === 0) continue;
    maxWeight += Math.max(...w);
    minWeight += Math.min(...w);
  }
  const weightRange = maxWeight - minWeight;
  if (weightRange <= 0) return 0.5;

  const vec = buildComparisonVector(rowA, rowB, fields);
  let total = 0;
  for (let k = 0; k < fields.length; k++) {
    const f = fields[k]!;
    const level = vec[k]!;
    const w = em.matchWeights[f.field];
    if (!w) continue;
    total += w[level] ?? 0;
  }
  return (total - minWeight) / weightRange;
}

// ---------------------------------------------------------------------------
// Fallback result for tiny datasets
// ---------------------------------------------------------------------------

function fallbackResult(mk: MatchkeyConfig): EMResult {
  const m: Record<string, number[]> = {};
  const u: Record<string, number[]> = {};
  const w: Record<string, number[]> = {};
  for (const f of mk.fields) {
    const n = fieldLevels(f);
    if (n === 2) {
      m[f.field] = [0.1, 0.9];
      u[f.field] = [0.9, 0.1];
      w[f.field] = [Math.log2(0.1 / 0.9), Math.log2(0.9 / 0.1)];
    } else if (n === 3) {
      m[f.field] = [0.05, 0.15, 0.8];
      u[f.field] = [0.8, 0.15, 0.05];
      w[f.field] = [
        Math.log2(0.05 / 0.8),
        Math.log2(0.15 / 0.15),
        Math.log2(0.8 / 0.05),
      ];
    } else {
      // Uniform fallback.
      const mv = new Array<number>(n).fill(1 / n);
      const uv = new Array<number>(n).fill(1 / n);
      m[f.field] = mv;
      u[f.field] = uv;
      w[f.field] = new Array<number>(n).fill(0);
    }
  }
  return {
    m,
    u,
    matchWeights: w,
    proportionMatched: 0.05,
    iterations: 0,
    converged: false,
  };
}
